Sources/SourceKitD/SourceKitD.swift

```diff
@@ -83,56 +96,26 @@ extension SourceKitD {
       }
       return handle
     } cancel: { handle in
-      api.cancel_request(handle)
+      if let handle {
+        api.cancel_request(handle)
+      }
     }
 
-    logResponse(sourcekitdResponse)
+    log(response: sourcekitdResponse)
 
     guard let dict = sourcekitdResponse.value else {
-      signposter.endInterval("sourcekitd-request", signposterState, "Error")
       if sourcekitdResponse.error == .connectionInterrupted {
-        let log = """
-          Request:
-          \(req.description)
-
-          File contents:
-          \(fileContents ?? "<nil>")
-          """
-        let chunks = splitLongMultilineMessage(message: log)
-        for (index, chunk) in chunks.enumerated() {
-          logger.fault(
-            """
-            sourcekitd crashed (\(index + 1)/\(chunks.count))
-            \(chunk)
-            """
-          )
-        }
+        log(crashedRequest: req, fileContents: fileContents)
       }
       throw sourcekitdResponse.error!
     }
 
-    signposter.endInterval("sourcekitd-request", signposterState, "Done")
     return dict
   }
-}
-
-private func logRequest(_ request: SKDRequestDictionary) {
-  logger.info(
-    """
-    Sending sourcekitd request:
-    \(request.forLogging)
-    """
-  )
-}
 
-private func logResponse(_ response: SKDResponse) {
-  logger.log(
-    level: (response.error == nil || response.error == .requestCancelled) ? .debug : .error,
-    """
-    Received sourcekitd response:
-    \(response.forLogging)
-    """
-  )
+  public func log(request: SKDRequestDictionary) {}
+  public func log(response: SKDResponse) {}
+  public func log(crashedRequest: SKDRequestDictionary, fileContents: String?) {}
```

Comment:
Is it worth having empty defaults?

Reply:
True, probably better to not have them and force adopters to make decisions about whether they support logging or not.